### PR TITLE
[Buttons] Creating animations for FAB appearance

### PR DESCRIPTION
**Pull request by romoore**
The Floating Action Button (FAB) is supposed to "pop-in" to a view when the
view appears. Adding some basic animation code to allow the FAB to be
animated-in and -out of views.

fab-animate

### CODE REVIEW

File: components/Buttons/src/MDCFloatingButton+Animation.m [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+@implementation MDCFloatingButton (Animation)
+
+- (void)grow:(BOOL)animated completion:(void (^_Nullable)(void))completion {
+  void (^growActions)() = ^{
+    self.transform = CGAffineTransformIdentity;
+  };
+  if (animated) {
+    [UIView animateWithDuration:kMDCFloatingButtonAnimationDuration
+                     animations:^{
+                       growActions();
```
Comment (willlarche) [inline components/Buttons/src/MDCFloatingButton+Animation.m]: Since the block is a void block, I'd just pass it to animations: instead of invoking it inside.

File: components/Buttons/src/MDCFloatingButton+Animation.m [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+@implementation MDCFloatingButton (Animation)
+
+- (void)grow:(BOOL)animated completion:(void (^_Nullable)(void))completion {
+  void (^growActions)() = ^{
+    self.transform = CGAffineTransformIdentity;
+  };
+  if (animated) {
+    [UIView animateWithDuration:kMDCFloatingButtonAnimationDuration
+                     animations:^{
```
Comment (willlarche) [inline components/Buttons/src/MDCFloatingButton+Animation.m]: This probably needs a curve / motion design. Did it get vetted by a motion designer?

File: components/Buttons/src/MDCFloatingButton+Animation.h [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+
+#import "MDCFloatingButton.h"
+
+@interface MDCFloatingButton (Animation)
+
+/**
+ Causes this button to return to its "normal" size.
```
Comment (willlarche) [inline components/Buttons/src/MDCFloatingButton+Animation.h]: I would love to have the actual diameter in this comment. It's GTK.

File: components/Buttons/examples/ButtonsTypicalUse.m [hunk 4]
```diff
@@ -20,17 +20,15 @@
...
 @implementation ButtonsTypicalUseViewController
 
 - (MDCButton *)buildCustomStrokedButton {
   MDCButton *button = [[MDCButton alloc] init];
-  [button setBackgroundColor:[UIColor clearColor]
-                    forState:UIControlStateNormal];
-  [button setTitleColor:[UIColor colorWithWhite:0.1 alpha:1.0]
-               forState:UIControlStateNormal];
+  [button setBackgroundColor:[UIColor clearColor] forState:UIControlStateNormal];
+  [button setTitleColor:[UIColor colorWithWhite:0.1 alpha:1.0] forState:UIControlStateNormal];
```
Comment (willlarche) [inline components/Buttons/examples/ButtonsTypicalUse.m]: You probably need to cast these floats to float with f. Build for 32 and see.

File: components/Buttons/examples/ButtonsTypicalUse.m [hunk 5]
```diff
@@ -119,33 +117,52 @@ - (void)viewDidLoad {
...
   ];
 
   [self setupExampleViews];
 }
 
 - (void)didTap:(id)sender {
   NSLog(@"%@ was tapped.", NSStringFromClass([sender class]));
+  if (sender == self.floatingButton) {
+    [self.floatingButton shrink:YES
+                               completion:^{
```
Comment (willlarche) [inline components/Buttons/examples/ButtonsTypicalUse.m]: This alignment looks a little off. I think it should be aligned to colons.

File: components/Buttons/examples/supplemental/ButtonsTypicalUseSupplemental.m [hunk 6]
```diff
@@ -89,12 +84,11 @@ - (void)viewWillLayoutSubviews {
     CGFloat colOffset = CGRectGetWidth(self.view.frame) / 4;
```
Comment (willlarche) [inline components/Buttons/examples/supplemental/ButtonsTypicalUseSupplemental.m]: I know this isn't your writing but I'd love to see 'col' not abbreviated.

**Review by willlarche — CHANGES_REQUESTED** [review]

File: components/Buttons/src/MDCFloatingButton+Animation.m (hunk 2 above)
Comment (romoore) [reply components/Buttons/src/MDCFloatingButton+Animation.m]: Nope, marked as "needing research" until I can get the correct motion metrics.

File: components/Buttons/src/MDCFloatingButton+Animation.m (hunk 1 above)
Comment (romoore) [reply components/Buttons/src/MDCFloatingButton+Animation.m]: Good idea! Done

File: components/Buttons/src/MDCFloatingButton+Animation.h (hunk 3 above)
Comment (romoore) [reply components/Buttons/src/MDCFloatingButton+Animation.h]: I'm a little wary of documenting the exact value (in case we change it) - how about "scaled to less than 0.1% of the original size"?

File: components/Buttons/examples/ButtonsTypicalUse.m (hunk 5 above)
Comment (romoore) [reply components/Buttons/examples/ButtonsTypicalUse.m]: Agreed, seems I can make it ever-so-slightly better.

File: components/Buttons/examples/supplemental/ButtonsTypicalUseSupplemental.m (hunk 6 above)
Comment (romoore) [reply components/Buttons/examples/supplemental/ButtonsTypicalUseSupplemental.m]: Agreed. Done.

File: components/Buttons/examples/ButtonsTypicalUse.m [hunk 7]
```diff
@@ -20,28 +20,26 @@
...
-  [button setTitleColor:[UIColor colorWithWhite:0.1 alpha:1.0]
-               forState:UIControlStateNormal];
-  button.inkColor = [UIColor colorWithWhite:0 alpha:0.06];
+  [button setBackgroundColor:[UIColor clearColor] forState:UIControlStateNormal];
+  [button setTitleColor:[UIColor colorWithWhite:0.1F alpha:1] forState:UIControlStateNormal];
+  button.inkColor = [UIColor colorWithWhite:0 alpha:0.06F];
   button.layer.borderWidth = 1;
   button.layer.borderColor = [UIColor blackColor].CGColor;
-  button.disabledAlpha = 0.38;
+  button.disabledAlpha = 0.38F;
```
Comment (ajsecord) [inline components/Buttons/examples/ButtonsTypicalUse.m]: Floats in our codebase use lowercase 'f'.

File: components/Buttons/src/MDCFloatingButton+Animation.h [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+ @param animated YES if the size change should be animated.
+ @param completion a completion block to call after the size change is complete.
+ */
+- (void)enter:(BOOL)animated completion:(void (^_Nullable)(void))completion;
```
Comment (ajsecord) [inline components/Buttons/src/MDCFloatingButton+Animation.h]: Please name this more descriptively—at the call site reading `enter` doesn't make it clear what is going on. Consider naming it for the actual animation, in case there are multiple kinds in the future.

Comment (ajsecord) [reply components/Buttons/src/MDCFloatingButton+Animation.h]: (and same for `exit`, obviously)

File: components/Buttons/src/MDCFloatingButton+Animation.m [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+    self.layer.transform = CATransform3DIdentity;
+    self.layer.opacity = 1;
+    self.imageView.layer.transform = CATransform3DIdentity;
+    [self.layer removeAnimationForKey:@"kMDCFloatingButtontransform"];
+    [self.layer removeAnimationForKey:@"kMDCFloatingButtonopacity"];
+    [self.imageView.layer removeAnimationForKey:@"kMDCFloatingButtontransform"];
+    if (completion) {
+      completion();
+    }
+    NSLog(@"%@", [self.layer animationKeys]);
```
Comment (ajsecord) [inline components/Buttons/src/MDCFloatingButton+Animation.m]: Remove all debugging logs.

File: components/Buttons/src/MDCFloatingButton+Animation.m [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+
+  [self.layer removeAllAnimations];
+  [self.imageView.layer removeAllAnimations];
+
+  if (animated) {
+    [CATransaction begin];
+    [CATransaction setDisableActions:YES];
+    [CATransaction setCompletionBlock:enterActions];
+
+    CABasicAnimation *overallScaleAnimation = [CABasicAnimation animationWithKeyPath:@"transform"];
```
Comment (ajsecord) [inline components/Buttons/src/MDCFloatingButton+Animation.m]: These animation blocks are suspiciously similar. Refactor into functions/methods.

File: components/Buttons/src/MDCFloatingButton+Animation.m [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+#endif
+    [self.layer addAnimation:opacityAnimation forKey:@"kMDCFloatingButtonopacityexit"];
+
+    [CATransaction commit];
+  } else {
+    exitActions();
+  }
+}
+
+#pragma mark - UIView overrides to handle non-identity transforms
```
Comment (ajsecord) [inline components/Buttons/src/MDCFloatingButton+Animation.m]: This needs more documentation—why is this required?

**Review by ajsecord — CHANGES_REQUESTED** [review]

File: components/Buttons/examples/ButtonsTypicalUse.m (hunk 4 above)
Comment (romoore) [reply components/Buttons/examples/ButtonsTypicalUse.m]: It didn't complain when I tried building, but I changed anyway.

File: components/Buttons/examples/ButtonsTypicalUse.m (hunk 7 above)
Comment (romoore) [reply components/Buttons/examples/ButtonsTypicalUse.m]: Done

File: components/Buttons/src/MDCFloatingButton+Animation.h (hunk 8 above)
Comment (romoore) [reply components/Buttons/src/MDCFloatingButton+Animation.h]: I was trying to align to how we named ink enter/exit.  How about "expand" and "collapse" to align with the wording on the [motion guidelines]( https://material.io/guidelines/motion/transforming-material.html)?

File: components/Buttons/src/MDCFloatingButton+Animation.m (hunk 9 above)
Comment (romoore) [reply components/Buttons/src/MDCFloatingButton+Animation.m]: Yep, this commit was a work-in-progress save.

File: components/Buttons/examples/ButtonsTypicalUse.m (hunk 7 above)
Comment (ianegordon) [reply components/Buttons/examples/ButtonsTypicalUse.m]: This is still capitalized in the PR on origin.
git push?

File: components/Buttons/src/MDCFloatingButton+Animation.m [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+ limitations under the License.
+ */
+
+#import "MDCFloatingButton+Animation.h"
+
+#if TARGET_IPHONE_SIMULATOR
+float UIAnimationDragCoefficient(void);  // Private API for simulator animation speed
+#endif
+
+static const CGFloat kMDCFloatingButtonTransformScale = 0.00001F;
```
Comment (ianegordon) [inline components/Buttons/src/MDCFloatingButton+Animation.m]: s/F/f/
Here and below.

File: components/Buttons/src/MDCFloatingButton+Animation.m [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+static const NSTimeInterval kMDCFloatingButtonExitIconDuration = 0.135F;
+static const NSTimeInterval kMDCFloatingButtonExitIconOffset = 0.000F;
+
+static const NSTimeInterval kMDCFloatingButtonOpacityDuration = 0.015F;
+static const NSTimeInterval kMDCFloatingButtonOpacityEnterOffset = 0.030F;
+static const NSTimeInterval kMDCFloatingButtonOpacityExitOffset = 0.150F;
+
+@implementation MDCFloatingButton (Animation)
+
+#if TARGET_IPHONE_SIMULATOR
```
Comment (ianegordon) [inline components/Buttons/src/MDCFloatingButton+Animation.m]: Add a comment why we need this?  And why only on the simulator?

File: components/Buttons/src/MDCFloatingButton+Animation.m (hunk 11 above)
Comment (ianegordon) [reply components/Buttons/src/MDCFloatingButton+Animation.m]: While I understand we can't use frame for layout when our transform isn't the identity, changing .frame is very concerning.
Let's discuss some alternatives.

**Review by ianegordon — CHANGES_REQUESTED** [review]

File: components/Buttons/src/MDCFloatingButton+Animation.m (hunk 10 above)
Comment (romoore) [reply components/Buttons/src/MDCFloatingButton+Animation.m]: Done

File: components/Buttons/src/MDCFloatingButton+Animation.m (hunk 11 above)
Comment (romoore) [reply components/Buttons/src/MDCFloatingButton+Animation.m]: Done

File: components/Buttons/src/MDCFloatingButton+Animation.m (hunk 13 above)
Comment (romoore) [reply components/Buttons/src/MDCFloatingButton+Animation.m]: For debugging animations  using the simulator's "Slow animations" option.

File: components/Buttons/src/MDCFloatingButton+Animation.m (hunk 11 above)
Comment (romoore) [reply components/Buttons/src/MDCFloatingButton+Animation.m]: I've tried adding a container view, but MDCButton uses a shadowLayer so we can't swap them out.  It might be possible to refactor MDCButton to have an MDCButtonContainerView with a shadowLayer, then have all of the subviews in there.  The button view itself would just be an empty/transparent rectangle.  I can spend a day on that approach to see how far I can get if you prefer.

File: components/Buttons/src/MDCFloatingButton+Animation.m [hunk 14]
```diff
@@ -0,0 +1,242 @@
...
+  if (fabs(beginOffset) > DBL_EPSILON) {
+    animation.beginTime = CACurrentMediaTime() + (beginOffset * [self fab_dragCoefficient]);
+  }
+#endif
+
+  return animation;
+}
+
+#if TARGET_IPHONE_SIMULATOR
++ (float)fab_dragCoefficient {
```
Comment (willlarche) [inline components/Buttons/src/MDCFloatingButton+Animation.m:80]: Tell me more about this.

**Review by willlarche — APPROVED** [review]

File: components/Buttons/src/MDCFloatingButton+Animation.m (hunk 14 above)
Comment (romoore) [reply components/Buttons/src/MDCFloatingButton+Animation.m:80]: The purpose is to slow-down our custom Core Animation animations to match those of UIViews.  In the simulator, turning on Slow Animation changes this value.  Since it's a private API we don't want it compiled into "real" targets.  I stole the code from the FlexibleHeaderView.

File: components/Buttons/src/MDCFloatingButton+Animation.m [hunk 15]
```diff
@@ -0,0 +1,242 @@
...
+ the frame will be modified. Apple is very clear that if a UIView's transform is not the identity
+ transform, then the value of the frame is undefined and should be ignored.
+
+ Unfortunately, much of our layout code assumes that frames are valid and relies on them instead of
+ the view's center and bounds for positioning information. As a result, this method will determine
+ what the frame would look like without the transform by using the center and bounds.
+
+ https://developer.apple.com/documentation/uikit/uiview/1622621-frame
+ */
+- (CGRect)frame {
```
Comment (ajsecord) [inline components/Buttons/src/MDCFloatingButton+Animation.m]: I think this is too much magic and will burn us at some point in the future. The `frame` property is pretty central and special. For example, the page you link to says

> Changing the frame rectangle automatically redisplays the receiver without invoking the draw(_:) method.

How would that interact with this? Are we sure that this will continue to work in the future? 

Let's chat and figure out a more maintainable solution.

**Review by ajsecord — CHANGES_REQUESTED** [review]

**Comment by romoore** [comment]
During an offline discussion with ajsecord@, we've decided to remove the frame manipulation code from the MDCFloatingButton and provide clear guidance and reference to Apple's docs about how the frame is affected by transforms.

File: components/Buttons/src/MDCFloatingButton+Animation.h [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+@interface MDCFloatingButton (Animation)
+
+/**
+ Expand this button to its unscaled (normal) size.
+
+ @param animated YES if the size change should be animated.
+ @param completion a completion block to call after the size change is complete.
+ */
+- (void)expand:(BOOL)animated completion:(void (^_Nullable)(void))completion;
+
```
Comment (ajsecord) [inline components/Buttons/src/MDCFloatingButton+Animation.h]: As discussed, please add docs about how this manipulates the transform and include a pointer to Apple's relevant docs.

**Review by ajsecord — CHANGES_REQUESTED** [review]

File: components/Buttons/src/MDCFloatingButton+Animation.h [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+/**
+ Collapses this button so that it becomes smaller than 0.1% of its normal size.
+
+ @param animated YES if the size change should be animated.
+ @param completion a completion block to call after the size change is complete.
+
+ @note This method will modify the transform property of the button. Apple's documentation about
```
Comment (ajsecord) [inline components/Buttons/src/MDCFloatingButton+Animation.h:40]: Copy this to `expand:completion:`.

**Review by ajsecord — APPROVED** [review]

**Review by samnm — APPROVED** [review]
